packages/wallets/src/features/accounts/modules/TaxInformation/hooks/useTaxInformation.ts

```diff
@@ -0,0 +1,97 @@
+    taxIdentificationNumber: THooks.AccountSettings['tax_identification_number'];
+    taxResidence: THooks.AccountSettings['tax_residence'];
+};
+
+const useTaxInformation = () => {
+    const { data: residenceList, isLoading, isSuccess: isResidenceListSuccess } = useResidenceList();
+    const { data: accountSettings, error, isSuccess: isAccountSettingsSuccess, update } = useSettings();
+    const [isSubmissionInitiated, setIsSubmissionInitiated] = useState(false);
+    const [isSubmitted, setIsSubmitted] = useState(false);
+
+    const countryCodeToPatternMapper = useMemo(() => {
+        const countryCodeToPatternMapping: Record<TCountryCode, string> = {};
+
+        if (isResidenceListSuccess) {
+            residenceList.forEach(residence => {
+                if (residence.value && !((residence.value as TCountryCode) in countryCodeToPatternMapping)) {
+                    countryCodeToPatternMapping[residence.value] = residence?.tin_format?.[0] ?? '';
+                }
+            });
+        }
+        return countryCodeToPatternMapping;
+    }, [isResidenceListSuccess, residenceList]);
+
+    const countryList = useMemo(() => {
+        return residenceList.map(residence => ({
+            text: residence.text as TCountryName,
+            value: residence.value as TCountryCode,
+        })) as TCountryList;
+    }, [residenceList]);
+
+    const initialValues = useMemo(
+        () =>
+            ({
+                accountOpeningReason: accountSettings.account_opening_reason,
+                citizenship: accountSettings.citizen,
+                placeOfBirth: accountSettings.place_of_birth,
+                taxIdentificationNumber: accountSettings.tax_identification_number ?? undefined,
+                taxResidence: accountSettings.tax_residence ?? undefined,
+            } as TTaxInformationValues),
+        [
+            accountSettings.account_opening_reason,
+            accountSettings.citizen,
+            accountSettings.place_of_birth,
+            accountSettings.tax_identification_number,
+            accountSettings.tax_residence,
+        ]
+    );
+
+    const onSubmit = (values: FormikValues | TTaxInformationValues) => {
+        if (
+            values &&
+            values.placeOfBirth &&
+            values.taxResidence &&
+            values.accountOpeningReason &&
+            values.taxIdentificationNumber
+        ) {
+            update({
+                account_opening_reason: values.accountOpeningReason,
+                citizen: values.citizenship,
+                place_of_birth: values.placeOfBirth,
+                tax_identification_number: values.taxIdentificationNumber,
+                tax_residence: values.taxResidence,
+            });
+            setIsSubmissionInitiated(true);
+        }
+    };
+
+    if (isSubmissionInitiated && isAccountSettingsSuccess) {
+        setIsSubmissionInitiated(false);
+        setIsSubmitted(true);
+    }
```

Comment:
mmm i would like to see it at least in useEffect 🤔

Reply:
I am working on @wojciech-deriv suggestion to implement it using async/await.